src/ucp/wireup/wireup_ep.c

```diff
@@ -219,30 +235,78 @@ static ucs_status_t ucp_wireup_ep_pending_add(uct_ep_h uct_ep,
 }
 
 static void
-ucp_wireup_ep_pending_purge(uct_ep_h uct_ep, uct_pending_purge_callback_t cb,
-                            void *arg)
+ucp_wireup_ep_pending_purge_cb(uct_pending_req_t *self, void *arg)
+{
+    ucp_wireup_ep_pending_purge_arg_t *purge_arg =
+        (ucp_wireup_ep_pending_purge_arg_t*)arg;
+    ucp_wireup_ep_t *wireup_ep                   = purge_arg->wireup_ep;
+    uct_ep_h uct_ep                              =
+        ucp_wireup_ep_get_msg_ep(wireup_ep);
+    ucp_request_t *wireup_proxy_req              =
+        ucs_container_of(self, ucp_request_t, send.uct);
+
+    /* do purging on AUX EP or on UCT EP if WIREUP is an owner of it */
+    if ((uct_ep == wireup_ep->aux_ep) || wireup_ep->super.is_owner) {
+        /* need to NULL the WIREUP EP in the WIREUP MSG proxy pending request
+         * to avoid dereferencing it when progressing prnding requests, since
+         * it will be destroyed */
+        wireup_proxy_req->send.proxy.wireup_ep = NULL;
+
+        purge_arg->wireup_msg_cb(self, purge_arg->wireup_msg_arg);
+        if (purge_arg->wireup_msg_cb != ucp_wireup_ep_pending_req_release) {
+            /* decrement the pending count, if it is not a request release
+             * callback */
+            ucs_atomic_sub32(&wireup_ep->pending_count, 1);
+        }
+    }
+}
+
+void
+ucp_wireup_ep_pending_purge_common(ucp_wireup_ep_t *wireup_ep,
+                                   uct_pending_purge_callback_t wireup_msg_cb,
+                                   void *wireup_msg_arg,
+                                   uct_pending_purge_callback_t user_msg_cb,
+                                   void *user_msg_arg)
```

Comment:
this is over complication
why not mostly keep the old flow and just pass flag whether to release wireup messages or call the purge_cb for them

Reply:
we have to set wireup_ep to the `NULL` for debug purposes (to not touch it in the WIREUP MSG pending progress function) and to check that wireup_ep is `NULL` and to not decrement the pending count in this case